src/riscv/lib/src/machine_state/page_cache/state.rs

```diff
@@ -155,17 +158,28 @@ impl<const PAGES: usize, CPE: CodePageEntry<MC, M>, MC: MemoryConfig, M: Manager
     }
 
     /// Invalidate a range of pages corresponding to the provided range of memory.
-    fn invalidate_pages(&mut self, addresses: std::ops::Range<u64>) {
-        let start_page = address_to_page_index(addresses.start);
-        let end_page = address_to_page_index(addresses.end.wrapping_sub(1));
-
-        // shortcut in-case of ranges out of bounds of memory
-        let end_page = end_page.min(self.pages.len().saturating_sub(1));
+    fn invalidate_pages(&mut self, pages: std::ops::RangeInclusive<u64>) {
+        for page_idx in pages.take_while(|idx| *idx < PAGES as u64) {
+            self.pages[page_idx as usize] = None;
+        }
+    }
+}
 
-        for page_idx in start_page..=end_page {
-            if let Some(entry) = self.pages.get_mut(page_idx) {
-                *entry = None;
-            }
+impl<const PAGES: usize, CPE, MC, M> MemoryGovernanceListener for PageCacheImpl<PAGES, CPE, MC, M>
+where
+    CPE: CodePageEntry<MC, M>,
+    MC: MemoryConfig,
+    M: ManagerBase,
+{
+    fn handle_permissions_update(
+        &mut self,
+        pages: std::ops::RangeInclusive<u64>,
+        permissions: memory::Permissions,
+    ) {
+        let needs_invalidation = permissions.can_write() || !permissions.can_exec();
```

Comment:
The doc-comment for `handle_permissions_update` states the purpose of the function in a generic sense, but in the impl block like here it is not clear why I want the handle to perform invalidation. 

Don't know what's better, a comment for why invalidation is computed this way or a doc-string on top of this function, but one of these should be present here imo.

Reply:
added a doc-comment [here](https://github.com/tezos/riscv-pvm/compare/fdc27d492e92be2dd50d8bbf43761ae76ab443fc..7ea21465fedbb4c74e1b35f38380812a097345ee)